In [1]:
import os
os.chdir('C:\\Users\\LENOVO\\Downloads')

In [2]:
import pandas as pd
#Load the data
file_path = 'null.csv'
data = pd.read_csv(file_path)
#Display the first few rows of the dataframe
data.head()

Month_Year Area Type              Borough_SNT                Area name  \
0  2023-04-01   Borough  Aviation Security(SO18)  Aviation Security(SO18)   
1  2023-04-01   Borough  Aviation Security(SO18)  Aviation Security(SO18)   
2  2023-04-01   Borough  Aviation Security(SO18)  Aviation Security(SO18)   
3  2023-04-01   Borough  Aviation Security(SO18)  Aviation Security(SO18)   
4  2023-04-01   Borough  Aviation Security(SO18)  Aviation Security(SO18)   

  Area code                         Offence Group  \
0      SO18             Arson and Criminal Damage   
1      SO18                         Drug Offences   
2      SO18                         Drug Offences   
3      SO18                         Drug Offences   
4      SO18  Miscellaneous Crimes Against Society   

                                    Offence Subgroup   Measure Financial Year  \
0                                    Criminal Damage  Offences        fy23-24   
1                                   Drug Trafficking  Offences        fy23-24   
2                                Possession of Drugs  Offences        fy23-24   
3                                Possession of Drugs  Outcomes        fy23-24   
4  Making, Supplying or Possessing Articles for u...  Offences        fy23-24   

   FY_FYIndex  Count  
0  fy23-24_07      3  
1  fy23-24_07      1  
2  fy23-24_07      4  
3  fy23-24_07      3  
4  fy23-24_07      2

In [3]:
#Convert 'Month_Year' to datetime format
data['Month_Year'] = pd.to_datetime(data['Month_Year'])
#Aggregate the data by 'Month_Year', summing up the 'Count' column
time_series_data = data.groupby('Month_Year')['Count'].sum().reset_index()
#Sort the data based on 'Month_Year'
time_series_data = time_series_data.sort_values('Month_Year')
#Display the aggregated time series
time_series_data.head()

Month_Year   Count
0 2023-04-01  169034
1 2023-05-01  185152
2 2023-06-01  191128
3 2023-07-01  188698

In [4]:
#Check the range of dates to confirm if any are missing
date_range = pd.date_range(start=time_series_data['Month_Year'].min(), 
                           end=time_series_data['Month_Year'].max(), 
                           freq='MS')  # 'MS' stands for Month Start frequency
#Identify if there are any missing months in the data
missing_dates = date_range.difference(time_series_data['Month_Year'])
missing_dates

DatetimeIndex([], dtype='datetime64[ns]', freq='MS')

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
#Splitting the data: 75% for training and 25% for testing
train_data, test_data = train_test_split(time_series_data, test_size=0.25, shuffle=False)
#Normalizing the data
scaler = MinMaxScaler(feature_range=(0, 1))
train_scaled = scaler.fit_transform(train_data['Count'].values.reshape(-1, 1))
test_scaled = scaler.transform(test_data['Count'].values.reshape(-1, 1))
# Reshaping the data to fit the RNN input requirements
# RNNs require input shape of the form [samples, time steps, features]
# Here, each sample is one month, and we have one feature - the count
#We have only one feature, so we reshape the data to [samples, time steps=1, features=1]
train_scaled = train_scaled.reshape((train_scaled.shape[0], 1, train_scaled.shape[1]))
test_scaled = test_scaled.reshape((test_scaled.shape[0], 1, test_scaled.shape[1]))
#Check the shapes of the processed data
(train_scaled.shape, test_scaled.shape)

((3, 1, 1), (1, 1, 1))

In [6]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
from tensorflow.keras.callbacks import EarlyStopping
#Define the LSTM model
model = Sequential()
model.add(LSTM(units=8, input_shape=(train_scaled.shape[1], train_scaled.shape[2])))
model.add(Dense(1))
model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])
#Early stopping callback to prevent overfitting
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
#Train the model with validation split
history = model.fit(
    train_scaled,
    train_scaled[:, :, 0],  # The target is the same as the input in this case
    epochs=200,
    batch_size=1,
    verbose=1,
    validation_split=0.2,  # Use part of the training data for validation
    callbacks=[early_stopping]
)
#Evaluate the model on training data
train_mae = model.evaluate(train_scaled, train_scaled[:, :, 0], verbose=0)[1]
#Evaluate the model on test data
test_mae = model.evaluate(test_scaled, test_scaled[:, :, 0], verbose=0)[1]
print(f'Train MAE: {train_mae}')
print(f'Test MAE: {test_mae}')

Epoch 1/200
2/2 [==============================] - 2s 586ms/step - loss: 0.1860 - mae: 0.3050 - val_loss: 0.7030 - val_mae: 0.8385
Epoch 2/200
2/2 [==============================] - 0s 40ms/step - loss: 0.1846 - mae: 0.3054 - val_loss: 0.6934 - val_mae: 0.8327
Epoch 3/200
2/2 [==============================] - 0s 30ms/step - loss: 0.1804 - mae: 0.3026 - val_loss: 0.6854 - val_mae: 0.8279
Epoch 4/200
2/2 [==============================] - 0s 41ms/step - loss: 0.1790 - mae: 0.3035 - val_loss: 0.6762 - val_mae: 0.8223
Epoch 5/200
2/2 [==============================] - 0s 33ms/step - loss: 0.1762 - mae: 0.3025 - val_loss: 0.6673 - val_mae: 0.8169
Epoch 6/200
2/2 [==============================] - 0s 13ms/step - loss: 0.1722 - mae: 0.2998 - val_loss: 0.6592 - val_mae: 0.8119
Epoch 7/200
2/2 [==============================] - 0s 31ms/step - loss: 0.1709 - mae: 0.3007 - val_loss: 0.6504 - val_mae: 0.8065
Epoch 8/200
2/2 [==============================] - 0s 31ms/step - loss: 0.1682 - mae: 0.2

2/2 [==============================] - 0s 31ms/step - loss: 0.0671 - mae: 0.2419 - val_loss: 0.2630 - val_mae: 0.5129
Epoch 64/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0651 - mae: 0.2392 - val_loss: 0.2582 - val_mae: 0.5082
Epoch 65/200
2/2 [==============================] - 0s 32ms/step - loss: 0.0641 - mae: 0.2381 - val_loss: 0.2535 - val_mae: 0.5035
Epoch 66/200
2/2 [==============================] - 0s 32ms/step - loss: 0.0639 - mae: 0.2384 - val_loss: 0.2482 - val_mae: 0.4982
Epoch 67/200
2/2 [==============================] - 0s 26ms/step - loss: 0.0621 - mae: 0.2358 - val_loss: 0.2437 - val_mae: 0.4936
Epoch 68/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0618 - mae: 0.2360 - val_loss: 0.2385 - val_mae: 0.4884
Epoch 69/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0601 - mae: 0.2335 - val_loss: 0.2342 - val_mae: 0.4839
Epoch 70/200
2/2 [==============================] - 0s 38ms/step - loss: 0.0592 - mae: 0.2324 - 

2/2 [==============================] - 0s 24ms/step - loss: 0.0309 - mae: 0.1743 - val_loss: 0.0798 - val_mae: 0.2824
Epoch 126/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0307 - mae: 0.1737 - val_loss: 0.0783 - val_mae: 0.2798
Epoch 127/200
2/2 [==============================] - 0s 19ms/step - loss: 0.0304 - mae: 0.1727 - val_loss: 0.0769 - val_mae: 0.2772
Epoch 128/200
2/2 [==============================] - 0s 29ms/step - loss: 0.0301 - mae: 0.1718 - val_loss: 0.0755 - val_mae: 0.2748
Epoch 129/200
2/2 [==============================] - 0s 29ms/step - loss: 0.0297 - mae: 0.1706 - val_loss: 0.0745 - val_mae: 0.2729
Epoch 130/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0295 - mae: 0.1700 - val_loss: 0.0731 - val_mae: 0.2705
Epoch 131/200
2/2 [==============================] - 0s 30ms/step - loss: 0.0292 - mae: 0.1688 - val_loss: 0.0722 - val_mae: 0.2686
Epoch 132/200
2/2 [==============================] - 0s 32ms/step - loss: 0.0290 - mae: 0.

Epoch 187/200
2/2 [==============================] - 0s 18ms/step - loss: 0.0164 - mae: 0.1227 - val_loss: 0.0303 - val_mae: 0.1742
Epoch 188/200
2/2 [==============================] - 0s 29ms/step - loss: 0.0162 - mae: 0.1219 - val_loss: 0.0299 - val_mae: 0.1730
Epoch 189/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0160 - mae: 0.1211 - val_loss: 0.0294 - val_mae: 0.1713
Epoch 190/200
2/2 [==============================] - 0s 30ms/step - loss: 0.0158 - mae: 0.1203 - val_loss: 0.0288 - val_mae: 0.1697
Epoch 191/200
2/2 [==============================] - 0s 31ms/step - loss: 0.0156 - mae: 0.1195 - val_loss: 0.0285 - val_mae: 0.1687
Epoch 192/200
2/2 [==============================] - 0s 30ms/step - loss: 0.0154 - mae: 0.1188 - val_loss: 0.0281 - val_mae: 0.1676
Epoch 193/200
2/2 [==============================] - 0s 22ms/step - loss: 0.0152 - mae: 0.1180 - val_loss: 0.0275 - val_mae: 0.1660
Epoch 194/200
2/2 [==============================] - 0s 54ms/step - loss: 0.

In [7]:
#Summary in form of a table
summary_results = {
    'Data Split': ['Training', 'Test'],
    'MAE': [train_mae, test_mae]
}
results_table = pd.DataFrame(summary_results)
#Display the results table
print(results_table)

  Data Split       MAE
0   Training  0.127053
1       Test  0.123598


The table displays the Mean Absolute Error (MAE) of the LSTM model on both the training and test data splits. The training data resulted in an MAE of approximately 0.127, while the model achieved a slightly better MAE of approximately 0.124 on the unseen test data, indicating that the model has a consistent performance on both datasets within a close margin.

The LSTM model's performance, as indicated by the Mean Absolute Error (MAE), shows that it has learned to predict the monthly count of offenses with a reasonable level of accuracy given the very limited dataset. The MAE of 0.127 for the training set and 0.124 for the test set suggests that the model is not overfitting, as the test error is slightly lower than the training error. This is a positive outcome, especially considering the challenges posed by the small amount of data available.